### ISSUE TITLE
Support negative step sizes for slicing

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#59786]
The [array API specification stipulates](https://data-apis.org/array-api/latest/API_specification/indexing.html#slice-syntax) that the slicing syntax needs to be equivalent to Pythons [`slice()`](https://docs.python.org/3/library/functions.html#slice) syntax. That includes negative step sizes that PyTorch does not support:

```python
import torch

t = torch.rand((3,))
t[::-1]
```

```
ValueError: step must be greater than zero
```

## Additional context

This was already proposed in #229, but finally dismissed in https://github.com/pytorch/pytorch/issues/229#issuecomment-548092342 in favor of `torch.flip()`. Note that this is only a substitution for a negative step size of 1 (`::-1`) rather than negative step sizes in general.

**Comment by rgommers** [comment]
Hmm that's a little surprising. Perhaps it's easier to implement now that there's no more TH?

Note that this is not advanced indexing (as the label says now), simple Python lists and tuples support it as well:
```python
>>> l = [1, 2, 3, 4, 5]
>>> l[::-2]
[5, 3, 1]
>>> x = np.arange(5)
>>> x[::-2]
array([4, 2, 0])
>>> t = torch.arange(5)
>>> t[::-2]
...
ValueError: step must be greater than zero

```

**Comment by SsnL** [comment]
It'll be weird unless we make it a view (rather than a copy)

**Comment by BeomyoungKim** [comment]
torch              1.12.1
torchaudio         0.12.1
torchvision        0.13.1

**Comment by BeomyoungKim** [comment]
How is it going? 
python 3.9.6
torch 1.12.1
torchaudio 0.12.1
torchvision 0.13.1
still get error ValueError: step must be greater than zero

**Comment by scaomath** [comment]
> ## Additional context
> This was already proposed in #229, but finally dismissed in [#229 (comment)](https://github.com/pytorch/pytorch/issues/229#issuecomment-548092342) in favor of `torch.flip()`. Note that this is only a substitution for a negative step size of 1 (`::-1`) rather than negative step sizes in general.
> 
> cc @mruberry @rgommers @pmeier

Just to add some context, `torch.flip()` is definitely not the most ideal answer, just to see the following two simple examples with bigger than 1 step size, or there are ellipsis:

### Example 1

```python
>>> x = np.arange(10)
>>> x[::-2]
array([9, 7, 5, 3, 1])

>>> y = torch.arange(10)
>>> y[::2].flip(dims=(0,))
tensor([8, 6, 4, 2, 0])

>>> y[1::2].flip(dims=(0,))
tensor([9, 7, 5, 3, 1])
```

For numpy the negative step size starts from the last index (convenient and intuitive), yet a positive step + `torch.flip()` needs a manual tweak when the length of that dimension modulo the step size is 0.

Now the worst nightmare is that this only happens when `len(x) % step_size == 0`, suppose we use the trick above
```python
>>> np.arange(10).reshape(2,5)[..., ::-2]
array([[4, 2, 0],
       [9, 7, 5]])
>>> y = torch.arange(10).view(2,5)
>>> y[...,1::2].flip(dims=(-1,))
tensor([[3, 1],
        [8, 6]])
>>> y[...,::2].flip(dims=(-1,))
tensor([[4, 2, 0],
        [9, 7, 5]])
```


### Example 2
When there are ellipsis and `newaxis`, the lack of negative step size invites the users to do more hard-coded indexing. To continue the example above

```
>>> x[..., ::-2, None]
array([[[4],
        [2],
        [0]],

       [[9],
        [7],
        [5]]])

>>> y[..., ::2, None].flip(dims=(-1,))
tensor([[[0],
         [2],
         [4]],

        [[5],
         [7],
         [9]]])
```

To replicate the numpy behavior, one has to do `y[..., ::2].flip(dims=(-1,))[..., None]` which is quite error prone.

**Comment by lucascolley** [comment]
FYI, this issue has come up during array API adoption in SciPy. This is currently the only blocker for [SciPy's Fast Hankel Transforms](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fht.html#scipy.fft.fht) to work with PyTorch Tensors.

**Comment by daniel-1388** [comment]
Exactly, using `xp.flip` will make **Fast Hankel Transforms** work on Pytorch Tensor. I've implemented that last week :).

**Comment by chansonzhang** [comment]
Can't use [::-1] to reverse a tensor is counter-intuitive. 
Examples listed by  @scaomath  is very practical.
